Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -187,3 +190,76 @@ lemma Derives.append_right {v w : List (Symbol T g.NT)}
   | tail _ last ih => exact ih.trans_produces <| last.append_right p
 
 end ContextFreeGrammar
+
+/-- Context-free languages are defined by context-free grammars. -/
+def Language.IsContextFree (L : Language T) : Prop :=
+  ∃ g : ContextFreeGrammar.{uT} T, g.language = L
+
+section closure_reversal
+
+private def reverseRule {N : Type uN} (r : ContextFreeRule T N) : ContextFreeRule T N :=
+  ⟨r.input, r.output.reverse⟩
+
+private def reverseGrammar (g : ContextFreeGrammar T) : ContextFreeGrammar T :=
+  ⟨g.NT, g.initial, g.rules.map reverseRule⟩
+
+private lemma reverseRule_reverseRule {N : Type uN} (r : ContextFreeRule T N) :
+    reverseRule (reverseRule r) = r := by
```

Comment:
You can use `Function.Involutive` here and `Function.Involutive.list_map` below.

Reply:
Something like that?